server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -562,7 +530,131 @@ public void onFailure(Exception e) {
+                    .append("] with total_size of [").append(new ByteSizeValue(recoveryResponse.phase1ExistingTotalSize))
+                    .append("]\n");
+                sb.append("   phase2: start took [").append(timeValueMillis(recoveryResponse.startTime)).append("]\n");
+                sb.append("         : recovered [").append(recoveryResponse.phase2Operations).append("]")
+                    .append(" transaction log operations")
+                    .append(", took [").append(timeValueMillis(recoveryResponse.phase2Time)).append("]")
+                    .append("\n");
+                logger.trace("{}", sb);
+            } else {
+                logger.debug("{} recovery done from [{}], took [{}]", request.shardId(), request.sourceNode(),
+                    recoveryTime);
+            }
+        }
+
+        @Override
+        public void handleException(TransportException e) {
+            onException(e);
+        }
+
+        private void onException(Exception e) {
+            if (logger.isTraceEnabled()) {
+                logger.trace(() -> new ParameterizedMessage(
+                    "[{}][{}] Got exception on recovery", request.shardId().getIndex().getName(),
+                    request.shardId().id()), e);
+            }
+            Throwable cause = ExceptionsHelper.unwrapCause(e);
+            if (cause instanceof CancellableThreads.ExecutionCancelledException) {
+                // this can also come from the source wrapped in a RemoteTransportException
+                onGoingRecoveries.failRecovery(recoveryId, new RecoveryFailedException(request,
+                    "source has canceled the recovery", cause), false);
+                return;
+            }
+            if (cause instanceof RecoveryEngineException) {
+                // unwrap an exception that was thrown as part of the recovery
+                cause = cause.getCause();
+            }
+            // do it twice, in case we have double transport exception
+            cause = ExceptionsHelper.unwrapCause(cause);
+            if (cause instanceof RecoveryEngineException) {
+                // unwrap an exception that was thrown as part of the recovery
+                cause = cause.getCause();
+            }
+
+            // here, we would add checks against exception that need to be retried (and not removeAndClean in this case)
+
+            if (cause instanceof IllegalIndexShardStateException || cause instanceof IndexNotFoundException ||
+                cause instanceof ShardNotFoundException) {
+                // if the target is not ready yet, retry
+                retryRecovery(
+                    recoveryId,
+                    "remote shard not ready",
+                    recoverySettings.retryDelayStateSync(),
+                    recoverySettings.activityTimeout());
+                return;
+            }
+
+            if (cause instanceof DelayRecoveryException || cause instanceof ResourceNotFoundException) {
```

Comment:
Can you add a comment here saying that the `ResourceNotFoundException` covers a failed `reestablishRecovery`? Given that `ResourceNotFoundException` is a super class for some exceptions (notably some that we handle earlier here), for example `RetentionLeaseNotFoundException`, that there's a risk that we might end up indefinitely retrying in case something's wrong with the recovery. I think I would prefer a more specialized exception or targeted approach here.